### ISSUE TITLE
Confusing error messages in generator-star

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#1949]
I started out trying to fix #1680, but I wasn't able to figure out what the expected behavior should be since the messages produced by `generator-star` are confusing, and the implementation seems incomplete.

For example:

```js
// "generator-star": [2, "end"]
function * foo() {}
```

Will fail with "Expected a space before *.", even though there is a space before the `*`.

More specifically, both of the following (theoretically invalid) test cases fail with zero errors:

```js
{
    code: "function*foo(){}",
    args: [2, "start"],
    ecmaFeatures: { generators: true },
    errors: [
        {
            message: "Expected a space after *.",
            type: "FunctionDeclaration"
        }
    ]
},
{
    code: "function*foo(){}",
    args: [2, "end"],
    ecmaFeatures: { generators: true },
    errors: [
        {
            message: "Expected a space before *.",
            type: "FunctionDeclaration"
        }
    ]
}
```

This one passes:

```js
{
    code: "function * foo(){}",
    args: [2, "start"],
    ecmaFeatures: { generators: true },
    errors: [
        {
            message: "Expected no space before *.",
            type: "FunctionDeclaration"
        }
    ]
}
```

But this one fails with the message "Expected a space before *.", even though a space is present:

```js
{
    code: "function * foo(){}",
    args: [2, "end"],
    ecmaFeatures: { generators: true },
    errors: [
        {
            message: "Expected no space after *.",
            type: "FunctionDeclaration"
        }
    ]
}
```

It's hard to understand what exactly each mode is attempting to enforce when `start` and `end` only check spacing on one side of the `*`. In fact, the only mode that checks spacing on both sides is `middle`, and it's undocumented.

I'd like to revamp this rule to check spacing on both sides and to report exactly the change necessary to fix the error. Then maybe it will be easier to figure out what should happen with #1680. What about changing the error messages to the following:

- `start` can produce "Unexpected space before *." and "Missing space after *."
- `end` can produce "Missing space before *." and "Unexpected space after *."
- `middle` can produce "Missing space before *.", "Missing space after *.", and "Missing space around *."

Thoughts?

**Comment by nzakas** [comment]
Yeah, this looks pretty confusing. I think what the rule really should be adding is "where do you want spaces with the star?" With the answers being "before" (start), " after " (end), " both", "neither".

By the way, go ahead and add labels when you file issues. Makes it easier for me to triage.

**Comment by xjamundx** [comment]
Sounds good to me! Why is the simple way always not quite good enough :)

**Comment by nzakas** [comment]
Question is, is this a big enough change to rename the rule to `generator-star-spacing`? Or is just changing the option names good enough?

**Comment by btmills** [comment]
I think `generator-star-spacing` is a clearer name, which matches better with the before/after/both/neither options. It'd also be easier to implement a new rule and deprecate `generator-star` in 1.0 rather than maintaining the existing behavior alongside the new one.

Working on this.
